src/FakeItEasy.Analyzer/NonVirtualSetupAnalyzer.cs

```diff
@@ -0,0 +1,108 @@
+            }
+
+            context.RegisterSyntaxNodeAction(AnalyzeMethod,
+                SyntaxKind.InvocationExpression);
+
+            context.RegisterSyntaxNodeAction(AnalyzeProperty,
+                SyntaxKind.SimpleMemberAccessExpression);
+        }
+
+        public static void AnalyzeMethod(SyntaxNodeAnalysisContext context)
+        {
+            //Method so exclude non-methods and interfaces
+
+            AnalyzeCall<InvocationExpressionSyntax>(context, x => NotMethod(x) || Interface(x));
+        }
+
+        public static void AnalyzeProperty(SyntaxNodeAnalysisContext context)
+        {
+            //Propery so exclude methods
+
+            AnalyzeCall<MemberAccessExpressionSyntax>(context, x => Method(x));
+        }
+
+        private static void AnalyzeCall<T>(SyntaxNodeAnalysisContext context, Func<SymbolInfo, bool> exclusions) where T : ExpressionSyntax
+        {
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(context.Node);
+
+            if (exclusions.Invoke(symbolInfo)) return;
+
+            var invocationExpression = (T)context.Node;
+
+            var invocationParent = invocationExpression
+                                    .Ancestors()
+                                    .OfType<InvocationExpressionSyntax>()
+                                    .FirstOrDefault();
+
+            if (!IsSetupInvocation(context, invocationParent))
+                return;
+
+            if (!symbolInfo.Symbol.IsVirtual)
+            {
+                var location = invocationExpression.GetLocation();
+
+                Diagnostic diagnostic = Diagnostic.Create(DiagnosticDefinitions.NonVirtualSetup, location, symbolInfo.Symbol.Name);
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private static bool IsSetupInvocation(SyntaxNodeAnalysisContext context, InvocationExpressionSyntax parent)
+        {
+            var methodSymbol = SymbolHelpers.GetCalledMethodSymbol(parent, context);
+
+            if (methodSymbol == null)
+            {
+                return false;
+            }
+
+            var methodFullName =
+                string.Concat(methodSymbol.ContainingType.GetFullName(), ".", methodSymbol.GetDecoratedName());
+
+            return DiagnosticsMap.ContainsKey(methodFullName);
+        }
+
+        private static ImmutableDictionary<string, DiagnosticDescriptor> CreateDiagnosticsMap()
+        {
+            var memberNames = DiagnosticSubjects.CallSpecMemberNames();
+
+            return memberNames.ToImmutableDictionary(name => name, name => DiagnosticDefinitions.NonVirtualSetup);
+        }
+
+        private static bool NotMethod(SymbolInfo x)
+        {
+            return x.Symbol?.Kind != SymbolKind.Method;
+        }
+
+        private static bool Method(SymbolInfo x)
+        {
+            return x.Symbol?.Kind == SymbolKind.Method;
+        }
+
+        private static bool Interface(SymbolInfo x)

```

Comment:
I find the naming of this method troubling. Unlike `NotMethod` and `Method` above, this method does not judge `x` (which is not a great parameter name—strive for something more meaningful)'s Interfaceness. Rather it checks to see if `x` were declared on an interface.
This really confused me when I was originally reading line 36, which I thought was saying to exclude items that were not methods or were interfaces (which aren't methods, and I couldn't understand why we'd be examining a symbol the could either be a method or an interface).